crates/iceberg/src/metrics.rs

```diff
@@ -0,0 +1,154 @@
+use crate::spec::SchemaId;
+
+/// This trait defines the API for reporting metrics of table operations.
+///
+/// Refer to the [Iceberg docs] for details.
+///
+/// [Iceberg docs]: https://iceberg.apache.org/docs/latest/metrics-reporting/
+#[async_trait]
+pub(crate) trait MetricsReporter: Debug + Send + Sync {
+    /// Indicates that an operation is done by reporting a MetricsReport.
+    ///
+    /// Any errors are expected to be handled internally.
+    async fn report(&self, report: MetricsReport);
+}
+
+/// An enum of all metrics reports.
+#[derive(Debug)]
+pub(crate) enum MetricsReport {
+    /// A Table Scan report that contains all relevant information from a Table Scan.
+    Scan {
+        table: TableIdent,
+        snapshot_id: i64,
+        schema_id: SchemaId,
+
+        /// If None, the scan is an unfiltered full table scan.
+        filter: Option<Arc<Predicate>>,
+
+        /// If None, the scan projects all fields.
+        // TODO: We could default to listing all field names in those cases: check what Java is doing.
+        projected_field_names: Option<Vec<String>>,
+        projected_field_ids: Arc<Vec<i32>>,
+
+        metrics: Arc<ScanMetrics>,
+        metadata: HashMap<String, String>,
+    },
+}
+
+/// Carries all metrics for a particular scan.
+#[derive(Debug)]
+pub(crate) struct ScanMetrics {
+    pub(crate) total_planning_duration: Duration,
+
+    // Manfiest-level metrics, computed by walking the snapshot's manifest list
+    // file entries and checking which manifests match the scan's predicates.
+    pub(crate) total_data_manifests: u32,
+    pub(crate) total_delete_manifests: u32,
+    pub(crate) skipped_data_manifests: u32,
+    pub(crate) skipped_delete_manifests: u32,
+    pub(crate) scanned_data_manifests: u32,
+    pub(crate) scanned_delete_manifests: u32,
+
+    // Data file-level metrics.
+    pub(crate) result_data_files: u32,
+    pub(crate) skipped_data_files: u32,
+    pub(crate) total_file_size_in_bytes: u64,
+
+    // Delete file-level metrics.
+    pub(crate) result_delete_files: u32,
+    pub(crate) skipped_delete_files: u32,
+    pub(crate) total_delete_file_size_in_bytes: u64,
+
+    pub(crate) indexed_delete_files: u32,
+    pub(crate) equality_delete_files: u32,
+    pub(crate) positional_delete_files: u32,
+}
+
+/// A reporter that logs the metrics to the console.
+#[derive(Clone, Debug)]
+pub(crate) struct LoggingMetricsReporter {}
+
+impl LoggingMetricsReporter {
+    pub(crate) fn new() -> Self {
+        Self {}
+    }
+}
+
+#[async_trait]
+impl MetricsReporter for LoggingMetricsReporter {
+    async fn report(&self, report: MetricsReport) {
+        match report {
+            MetricsReport::Scan {
+                table,
+                snapshot_id,
+                schema_id,
+                filter,
+                projected_field_names,
+                projected_field_ids,
+                metrics,
+                metadata,
+            } => {
+                info!(
```

Comment:
I don't think it's a good idea to use debug-formatted values here. I was struggling a lot to use the `tracing` API, and this is the best I could come up with so far.
I didn't really want to serialize the struct into a json, nor did I know how to implement `fmt::Display` for values such that they make sense across tracing subscribers.
Any suggestions welcome!